sstables/sstable_set.hh

```diff
@@ -111,7 +111,7 @@ public:
         schema_ptr,
         reader_permit,
         utils::estimated_histogram&,
-        const dht::partition_range&, // must be singular
+        const dht::ring_position&,
```

Comment:
I think that partition_key or decorated_key is more accurate.

ring_position doesn't necessarily denote a partition. It can denote a token.

Reply:
(ring_position denotes a boundary, not a point)

Reply:
We discussed this with @denesb on https://github.com/scylladb/scylla/pull/7437#discussion_r506067353:

the problem is that `sstable_set::select` (called in this function) takes a `partition_range`, which is constructed from `ring_position`s. Therefore I need a `ring_position` inside this function. I could construct it from passed-in `partition_key` or `decorated_key`, but that would involve a copy, and I want to avoid unnecessary copies here (the caller of the function would first extract the key from `ring_position`, then we would again copy it into a new `ring_position` --- lots of unnecessary work).

Passing `ring_position` instead of `partition_range` is still an improvement.

However, if you think that the copy is not so bad I could do it (in the context it's not really much, we're reading from disk here after all).

The right solution would probably be to modify `select` to take some kind of new type of `partition_range` that uses `ring_position_view`s. Constructing a `ring_position_view` from a `decorated_key` is cheap. However, we don't have that kind of `partition_range` class, so I'd have to write it.

Reply:
I think we can leave it as a `dht::ring_position`. The best solution would be to squash this with range version and decide on the reader to create internally, based on whether the range is singular or not.

In fact we should have a single factory function, which either leaves the wrapping with `restricting_reader` to the caller, or we add a flag `wait_for_admission` and do it if this is `wait_for_admission::yes`. We can do this in a separate PR.

Reply:
ok.

Reply:
For now I just added a comment:
```
+        const dht::ring_position&, // must contain a key
```